### ISSUE TITLE
Is it possible for kubelet to pass pod name and namespace to device plugin?

### ISSUE THREAD

**Issue by Garrybest** [kubernetes/kubernetes#119722]
### What would you like to be added?

Now device-plugin interacts with kubelet by a couple of gRPC interfaces. Here is the interface definition.

```go
// DevicePluginClient is the client API for DevicePlugin service.
//
// For semantics around ctx use and closing/ending streaming RPCs, please refer to https://godoc.org/google.golang.org/grpc#ClientConn.NewStream.
type DevicePluginClient interface {
	// GetDevicePluginOptions returns options to be communicated with Device
	// Manager
	GetDevicePluginOptions(ctx context.Context, in *Empty, opts ...grpc.CallOption) (*DevicePluginOptions, error)
	// ListAndWatch returns a stream of List of Devices
	// Whenever a Device state change or a Device disappears, ListAndWatch
	// returns the new list
	ListAndWatch(ctx context.Context, in *Empty, opts ...grpc.CallOption) (DevicePlugin_ListAndWatchClient, error)
	// GetPreferredAllocation returns a preferred set of devices to allocate
	// from a list of available ones. The resulting preferred allocation is not
	// guaranteed to be the allocation ultimately performed by the
	// devicemanager. It is only designed to help the devicemanager make a more
	// informed allocation decision when possible.
	GetPreferredAllocation(ctx context.Context, in *PreferredAllocationRequest, opts ...grpc.CallOption) (*PreferredAllocationResponse, error)
	// Allocate is called during container creation so that the Device
	// Plugin can run device specific operations and instruct Kubelet
	// of the steps to make the Device available in the container
	Allocate(ctx context.Context, in *AllocateRequest, opts ...grpc.CallOption) (*AllocateResponse, error)
	// PreStartContainer is called, if indicated by Device Plugin during registeration phase,
	// before each container start. Device plugin can run device specific operations
	// such as resetting the device before making devices available to the container
	PreStartContainer(ctx context.Context, in *PreStartContainerRequest, opts ...grpc.CallOption) (*PreStartContainerResponse, error)
}
```

We can see that methods `GetPreferredAllocation`, `Allocate` and `PreStartContainer` are used for device allocation and environment configuration.

However, the input parameters do not contain any information about **the pods and containers**, device-plugin have no idea about which pod or container is now being configured by kubelet, it only knows the **device IDs**.

Fortunately, these methods have a `ctx context.Context` as an input. I would like to add some information in this context when kubelet call device-plugins so that device plugins could know which pod or container is respective with this request. These information could make more extensibility for Kubernetes as device plugins could know more about which pod or container is relevant. **We don't need to modify the gRPC interface as well**.

### Why is this needed?

It is not enough for some device plugins to have only device IDs when receiving a gRPC request from kubelet. Taken method interface `Allocate` as an example, device plugins only know the device IDs.

```go
type DevicePluginClient interface {
        ......
	// Allocate is called during container creation so that the Device
	// Plugin can run device specific operations and instruct Kubelet
	// of the steps to make the Device available in the container
	Allocate(ctx context.Context, in *AllocateRequest, opts ...grpc.CallOption) (*AllocateResponse, error)
	......
}

type AllocateRequest struct {
	ContainerRequests    []*ContainerAllocateRequest `protobuf:"bytes,1,rep,name=container_requests,json=containerRequests,proto3" json:"container_requests,omitempty"`
	XXX_NoUnkeyedLiteral struct{}                    `json:"-"`
	XXX_sizecache        int32                       `json:"-"`
}

type ContainerAllocateRequest struct {
	DevicesIDs           []string `protobuf:"bytes,1,rep,name=devices_ids,json=devicesIds,proto3" json:"devices_ids,omitempty"`
	XXX_NoUnkeyedLiteral struct{} `json:"-"`
	XXX_sizecache        int32    `json:"-"`
}
```

However, some of our device plugins may need more information about this request. Here is some user stories.

1. I'm now developing a new network device plugin. When the plugin receives a `Allocate` request from kubelet, **it needs the pod name and namespace so it could retrieve the Pod from kube-apiserver or kubelet read-only port**. Our plugin may add some different configurations, like **different Envs or Mounts**, about this Pod based on the field `pod.spec.hostNetwork`.
2. I'm now developing a custom scheduler framework plugin so that scheduler could allocate devices for a Pod by adding device IDs in the pod annotations after scheduling. Our device pluin could know the fine-grained device scheduling result from pod annotations If device plugin could know the pod name and namespace.

In conclusion, I think pod base information, i.e. name and namespace could be passed to device plugins in order to elevate extensibility of device plugins while this patch does not need to make any modifications for device plugin gRPC interface. We only need to call a gRPC context method like `metadata.AppendToOutgoingContext(ctx, "pod", podNamespacedKey)` [here](https://github.com/kubernetes/kubernetes/blob/40741681a24a5acf9361ec74b97e887a9da3e3e4/pkg/kubelet/cm/devicemanager/endpoint.go#L85-L123).

**Comment by ffromani** [comment]
> ### What would you like to be added?
> 
> Now device-plugin interacts with kubelet by a couple of gRPC interfaces. Here is the interface definition.
> 
> ```go
> // DevicePluginClient is the client API for DevicePlugin service.
> //
> // For semantics around ctx use and closing/ending streaming RPCs, please refer to https://godoc.org/google.golang.org/grpc#ClientConn.NewStream.
> type DevicePluginClient interface {
> 	// GetDevicePluginOptions returns options to be communicated with Device
> 	// Manager
> 	GetDevicePluginOptions(ctx context.Context, in *Empty, opts ...grpc.CallOption) (*DevicePluginOptions, error)
> 	// ListAndWatch returns a stream of List of Devices
> 	// Whenever a Device state change or a Device disappears, ListAndWatch
> 	// returns the new list
> 	ListAndWatch(ctx context.Context, in *Empty, opts ...grpc.CallOption) (DevicePlugin_ListAndWatchClient, error)
> 	// GetPreferredAllocation returns a preferred set of devices to allocate
> 	// from a list of available ones. The resulting preferred allocation is not
> 	// guaranteed to be the allocation ultimately performed by the
> 	// devicemanager. It is only designed to help the devicemanager make a more
> 	// informed allocation decision when possible.
> 	GetPreferredAllocation(ctx context.Context, in *PreferredAllocationRequest, opts ...grpc.CallOption) (*PreferredAllocationResponse, error)
> 	// Allocate is called during container creation so that the Device
> 	// Plugin can run device specific operations and instruct Kubelet
> 	// of the steps to make the Device available in the container
> 	Allocate(ctx context.Context, in *AllocateRequest, opts ...grpc.CallOption) (*AllocateResponse, error)
> 	// PreStartContainer is called, if indicated by Device Plugin during registeration phase,
> 	// before each container start. Device plugin can run device specific operations
> 	// such as resetting the device before making devices available to the container
> 	PreStartContainer(ctx context.Context, in *PreStartContainerRequest, opts ...grpc.CallOption) (*PreStartContainerResponse, error)
> }
> ```
> 
> We can see that methods `GetPreferredAllocation`, `Allocate` and `PreStartContainer` are used for device allocation and environment configuration.
> 
> However, the input parameters do not contain any information about **the pods and containers**, device-plugin have no idea about which pod or container is now being configured by kubelet, it only knows the **device IDs**.
> 
> Fortunately, these methods have a `ctx context.Context` as an input. I would like to add some information in this context when kubelet call device-plugins so that device plugins could know which pod or container is respective with this request. These information could make more extensibility for Kubernetes as device plugins could know more about which pod or container is relevant. **We don't need to modify the gRPC interface as well**.
> ### Why is this needed?
> 
> It is not enough for some device plugins to have only device IDs when receiving a gRPC request from kubelet. Taken method interface `Allocate` as an example, device plugins only know the device IDs.
> 
> ```go
> type DevicePluginClient interface {
>         ......
> 	// Allocate is called during container creation so that the Device
> 	// Plugin can run device specific operations and instruct Kubelet
> 	// of the steps to make the Device available in the container
> 	Allocate(ctx context.Context, in *AllocateRequest, opts ...grpc.CallOption) (*AllocateResponse, error)
> 	......
> }
> 
> type AllocateRequest struct {
> 	ContainerRequests    []*ContainerAllocateRequest `protobuf:"bytes,1,rep,name=container_requests,json=containerRequests,proto3" json:"container_requests,omitempty"`
> 	XXX_NoUnkeyedLiteral struct{}                    `json:"-"`
> 	XXX_sizecache        int32                       `json:"-"`
> }
> 
> type ContainerAllocateRequest struct {
> 	DevicesIDs           []string `protobuf:"bytes,1,rep,name=devices_ids,json=devicesIds,proto3" json:"devices_ids,omitempty"`
> 	XXX_NoUnkeyedLiteral struct{} `json:"-"`
> 	XXX_sizecache        int32    `json:"-"`
> }
> ```
> 
> However, some of our device plugins may need more information about this request. Here is some user stories.
> 
>     1. I'm now developing a new network device plugin. When the plugin receives a `Allocate` request from kubelet, **it needs the pod name and namespace so it could retrieve the Pod from kube-apiserver or kubelet read-only port**. Our plugin may add some different configurations, like **different Envs or Mounts**, about this Pod based on the field `pod.spec.hostNetwork`.
> 
>     2. I'm now developing a custom scheduler framework plugin so that scheduler could allocate devices for a Pod by adding device IDs in the pod annotations after scheduling. Our device pluin could know the fine-grained device scheduling result from pod annotations If device plugin could know the pod name and namespace.
> 
> 
> In conclusion, I think pod base information, i.e. name and namespace could be passed to device plugins in order to elevate extensibility of device plugins while this patch does not need to make any modifications for device plugin gRPC interface. We only need to call a gRPC context method like `metadata.AppendToOutgoingContext(ctx, "pod", podNamespacedKey)` [here](https://github.com/kubernetes/kubernetes/blob/40741681a24a5acf9361ec74b97e887a9da3e3e4/pkg/kubelet/cm/devicemanager/endpoint.go#L85-L123).

Hi, I think this is KEP-worthy content and exceeds the scope of an issue. I for myself have mixed feelings about letting a device plugin know which pod is going to serve, even though I see the practical upside of this approach the design looks too coupled at first glance and somehow suspicious. I think it would best to engage with SIG-node to present the usecases and look for the best outcome.

For example, would [Dynamic Resource Allocation - DRA](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/3063-dynamic-resource-allocation) perhaps better serve your usecases?

**Comment by Garrybest** [comment]
Hi @ffromani, I agree with you. I'm worried about increasing coupling of device plugin and kubelet as well. So for me, I try to let this information deliverd as transparent as possible. For a gRPC call, delivery this information by context is kind of like an 'annotation' style, **so it's implicit**. Some device plugins could retrieve this information if necessary while mostly they don't need.

> For example, would [Dynamic Resource Allocation - DRA](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/3063-dynamic-resource-allocation) perhaps better serve your usecases?

I think it may be, but a little bit heavy. We now only use device plugin framework as a light way.

**Comment by hfingler** [comment]
Hi @Garrybest !  Did you find an alternative solution for this issue? I'm writing a plugin that needs an annotation or any other pod information to make a more informed decision, but I see nothing I can use in the API. Thanks!

**Comment by Garrybest** [comment]
> Hi @Garrybest ! Did you find an alternative solution for this issue? I'm writing a plugin that needs an annotation or any other pod information to make a more informed decision, but I see nothing I can use in the API. Thanks!

Hi @hfingler, If you use containerd, you could try to use CDI and NRI to solve this problem, I think these hooks could get Pod annotations.

**Comment by aojea** [comment]
NRI and DRA gives you all what you need https://github.com/aojea/kubernetes-network-driver?tab=readme-ov-file#anatomy-of-a-networking-dra-driver
